### ISSUE TITLE
Differentiation rule for 'create_token'

### ISSUE THREAD

**Issue by PhilipVinc** [google/jax#4292]
Over on `mpi4jax` we use `create_token` to force a sequential dependence on MPI operations.
I would like to compute gradients of functions that call MPI operations (the use case TLDR is to use in jax linear solver custom distributed matrix multiplications).

So to compute the gradient I need to differentiate `create_token`.

As the derivative does not really make sense here, it would be possible to return a strong zero.

Does this make sense to you?

**Comment by PhilipVinc** [comment]
Something along the lines of 

```python
from jax.lax import create_token, zeros_like_array
from jax.lax.lax import create_token_p
from jax.interpreters import ad

def create_token_value_and_jvp(in_args, tan_args):
    x, = in_args
    res = create_token(x)
    jvp = zeros_like_array(x)
    return (res, jvp)

ad.primitive_jvps[create_token_p] = create_token_value_and_jvp
```

**Comment by froystig** [comment]
As you say, there isn't a derivative here, so we shouldn't define a false one. JAX allows differentiation with respect to only some function inputs. At the user level, this corresponds to [passing `argnums` to functions such as `grad`](https://jax.readthedocs.io/en/latest/jax.html#jax.grad). Does this resolve things within your use case?

**Comment by PhilipVinc** [comment]
this does not resolve our issue.

The function I want to differentiate is:
```python
def Allreduce(x, op, comm=_MPI.COMM_WORLD, token=None):
    """
    if token is None:
        token = jax.lax.create_token(x)

   result, new_token = mpi_allreduce_p.bind(x, token, op=op, comm=comm)
   return result, new_token
```

(token is used to enforce a data dependency, because jax might reorder MPI operations).


We have implemented gradients for `mpi_allreduce_p`. If an user calls Allreduce by passing token kwarg a gradient can correctly be computed.

```python
>>> arr = jnp.ones((3, 2))
>>> token = jax.lax.create_token(arr)
>>> res, grad = jax.value_and_grad(lambda x: mpi4jax.Allreduce(x, op=MPI.SUM, token=token)[0].sum())(Sto arrivando!)
>>> res, grad = jax.value_and_grad(lambda x: Allreduce(x, op=MPI.SUM, token=token)[0].sum())(arr)
>>> res
DeviceArray(6., dtype=float32)
>>> grad
DeviceArray([[1., 1.],
             [1., 1.],
             [1., 1.]], dtype=float32)
```

However, if I don't pass in the token, even with `argnums` jax will still attempt to differentiate through the jax primitive `create_token`:

```python
>>> res, grad = jax.value_and_grad(lambda x: Allreduce(x, op=MPI.SUM)[0].sum(), argnums=0)(arr)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 1, in <lambda>
  File "/Users/filippovicentini/Dropbox/Ricerca/Codes/Python/mpi4jax/mpi4jax/collective_ops/allreduce.py", line 52, in Allreduce
    token = jax.lax.create_token(x)
  File "~/pythonenvs/mpi4jax_env/lib/python3.7/site-packages/jax/lax/lax.py", line 5703, in create_token
    return create_token_p.bind(x)
jax.traceback_util.FilteredStackTrace: NotImplementedError: Differentiation rule for 'create_token' not implemented

The stack trace above excludes JAX-internal frames.
The following is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "~/pythonenvs/mpi4jax_env/lib/python3.7/site-packages/jax/traceback_util.py", line 136, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "~/pythonenvs/mpi4jax_env/lib/python3.7/site-packages/jax/api.py", line 811, in value_and_grad_f
    ans, vjp_py = _vjp(f_partial, *dyn_args)
  File "~/pythonenvs/mpi4jax_env/lib/python3.7/site-packages/jax/api.py", line 1856, in _vjp
    out_primal, out_vjp = ad.vjp(flat_fun, primals_flat)
  File "~/pythonenvs/mpi4jax_env/lib/python3.7/site-packages/jax/interpreters/ad.py", line 110, in vjp
    out_primals, pvals, jaxpr, consts = linearize(traceable, *primals)
  File "~/pythonenvs/mpi4jax_env/lib/python3.7/site-packages/jax/interpreters/ad.py", line 97, in linearize
    jaxpr, out_pvals, consts = pe.trace_to_jaxpr(jvpfun_flat, in_pvals)
  File "~/pythonenvs/mpi4jax_env/lib/python3.7/site-packages/jax/interpreters/partial_eval.py", line 1095, in trace_to_jaxpr
    jaxpr, (out_pvals, consts, env) = fun.call_wrapped(pvals)
  File "~/pythonenvs/mpi4jax_env/lib/python3.7/site-packages/jax/linear_util.py", line 150, in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
  File "<stdin>", line 1, in <lambda>
  File "/Users/filippovicentini/Dropbox/Ricerca/Codes/Python/mpi4jax/mpi4jax/collective_ops/allreduce.py", line 52, in Allreduce
    token = create_token_jax(x)
  File "~/pythonenvs/mpi4jax_env/lib/python3.7/site-packages/jax/lax/lax.py", line 5703, in create_token
    return create_token_p.bind(x)
  File "~/pythonenvs/mpi4jax_env/lib/python3.7/site-packages/jax/core.py", line 1571, in bind
    out = top_trace.process_primitive(self, tracers, params)
  File "~/pythonenvs/mpi4jax_env/lib/python3.7/site-packages/jax/interpreters/ad.py", line 254, in process_primitive
    raise NotImplementedError(msg)
NotImplementedError: Differentiation rule for 'create_token' not implemented

**Comment by froystig** [comment]
It's helpful to see the code example now, and in particular the token's dependence on `x`. If you replace the statement `token = jax.lax.create_token(x)` with:

```python
token = jax.lax.create_token(jax.lax.stop_gradient(x))
```

I'd expect things to work as you intend. Do they?

**Comment by PhilipVinc** [comment]
yes! thanks a lot. That indeed solves the issue.
I had not thought about that...

**Comment by froystig** [comment]
As of #4330, you should no longer need to make that `stop_gradient` call yourself.

**Comment by PhilipVinc** [comment]
@froystig Sorry to reopen this issue, but I recently came across the need to have a `linear_transpose` rule for `create_token`.

The rationale is similar to what I had argued in the original post: we compute the transposition of some MPI reduction operation.
While now `create_token` includes `stop_gradient`, `stop_gradient` does not stop the transposition so the issue arises.

In general i'd argue that stop_gradient should not stop the transposition. 
Moreover, for this operation, the transposition could either be 'stopped' (by a `stop_transpose` function) or simply return an hard zero?

@cc inailuig
